src/primitives.jl

```diff
@@ -182,11 +195,18 @@ unwrapped `Nullable` entries. If no nulls are present, this is a no-op
 and `X` is returned.
 """
 function dropnull!{T}(X::AbstractVector{T})                 # -> AbstractVector
-    if !(Nullable <: T) && !(T <: Nullable)
+    u = isa(T, Union)
+    if !u && !(Nullable <: T) && !(T <: Nullable)
+        return X
+    elseif u && !any(S -> S <: Nullable || Nullable <: S, _uniontypes(T))
         return X
     else
         deleteat!(X, find(isnull, X))
-        res = similar(X, eltype(T))
+        if u
+            res = similar(X, Union{filter(t -> !(t <: Nullable), _uniontypes(T))...})
```

Comment:
You shouldn't filter out nullable types, since that means you won't be able to store some values in the result, e.g. if the input was `Union{Int, Nullable{String}}`. Need to call `eltype` instead. This needs to be tested. Also use upper case letters like `U` for types.

This should also go to a separate function, you really want to avoid any code duplication, especially for this kind of tricky checks.